src/main/java/org/opentripplanner/middleware/controllers/api/ApiUserController.java

```diff
@@ -1,28 +1,157 @@
+        ApiUser targetUser = getApiUser(req);
+        Auth0UserProfile requestingUser = Auth0Connection.getUserFromRequest(req);
+        String usagePlanId = req.queryParamOrDefault("usagePlanId", DEFAULT_USAGE_PLAN_ID);
+        // If requester is not an admin user, force the usage plan ID to the default and enforce key limit. A non-admin
+        // user should not be able to create an API key for any usage plan.
+        if (!isUserAdmin(requestingUser)) {
+            usagePlanId = DEFAULT_USAGE_PLAN_ID;
+            if (targetUser.apiKeys.size() >= API_KEY_LIMIT_PER_USER) {
+                logMessageAndHalt(req, HttpStatus.BAD_REQUEST_400, "User has reached API key limit.");
+            }
+        }
+        // FIXME Should an Api user be limited to one api key per usage plan (and perhaps stage)?
+        CreateApiKeyResult apiKeyResult = ApiGatewayUtils.createApiKey(targetUser.id, usagePlanId);
+        if (apiKeyResult == null || apiKeyResult.getId() == null) {
+            logMessageAndHalt(req,
+                HttpStatus.INTERNAL_SERVER_ERROR_500,
+                String.format("Unable to get AWS API key for user id (%s) and usage plan id (%s)", targetUser.id, usagePlanId),
+                null
+            );
+            return null;
+        }
+        // Add new API key to user and persist
+        targetUser.apiKeys.add(new ApiKey(apiKeyResult));
+        Persistence.apiUsers.replace(targetUser.id, targetUser);
+        return Persistence.apiUsers.getById(targetUser.id);
+    }
+
+    /**
+     * Delete an api key from a given user's list of api keys (if present) and from AWS api gateway.
+     */
+    private ApiUser deleteApiKeyForApiUser(Request req, Response res) {
+        ApiUser targetUser = getApiUser(req);
+        String apiKeyId = HttpUtils.getRequiredParamFromRequest(req, "apiKeyId");
+        if (apiKeyId == null) {
+            logMessageAndHalt(req,
+                HttpStatus.BAD_REQUEST_400,
+                "An api key id is required",
+                null);
+        }
+        if (!userHasKey(targetUser, apiKeyId)) {
+            logMessageAndHalt(req,
+                HttpStatus.NOT_FOUND_404,
+                String.format("User id (%s) does not have expected api key id (%s)", targetUser.id, apiKeyId),
+                null);
+        }
+
+        // Delete api key from AWS.
+        boolean success = ApiGatewayUtils.deleteApiKeys(Collections.singletonList(new ApiKey(apiKeyId)));
+        if (success) {
+            // Delete api key from user and persist
+            targetUser.apiKeys.removeIf(apiKey -> apiKeyId.equals(apiKey.id));
+            Persistence.apiUsers.replace(targetUser.id, targetUser);
+            return Persistence.apiUsers.getById(targetUser.id);
+        } else {
+            // Throw halt if API key deletion failed.
+            logMessageAndHalt(req, HttpStatus.INTERNAL_SERVER_ERROR_500, "Unknown error deleting API key.");
```

Comment:
If you attempt to delete a key that doesn't exist [anymore] in AWS, right now that throws a 500 error. Should it still delete that entry from Mongo? @evansiroky what do you think?

Reply:
@binh-dam-ibigroup I have refactored the `ApiGatewayUtils.deleteApiKeys(List<ApiKey> apiKeys)` to `ApiGatewayUtils.deleteApiKey(ApiKey apiKey)`. The older version had a return in the middle of a loop so would have only deleted one api key anyway! The newer version returns true even if the key is not found by AWS, the assumption being it has already been removed. False is still returned if any other exceptions occur.